### DISCUSSION TITLE
Inject symfony container / services

### DISCUSSION THREAD

**Discussion by darthf1** [phpbench/phpbench#995] [General]
Hi!

Apologies, this is more of a question than a report. If / when you enable Github discussions I'll happily move my question there.

I would like to start creating benchmarks for my code. To create these, I need to inject several services. I found the possibility to define services via the [extension](https://phpbench.readthedocs.io/en/latest/customisation/extension.html#create-the-dependency-injection-extension), however this means I need to recreate a lot of my configuration already defined in my symfony app, including configuration done by third party code in my vendor folder.

Is it possible to somehow inject the symfony container into my benchmarks, or register all services in the symfony container via the phpbench extension? I could not find anything on the subject, so I'm not sure this is a feasible / desirable request.

If it proves possible, I can give it a go and do a documentation PR. I see the following options. Maybe, in the extension, I can load the container via the XML file, the same as phpstan does in https://github.com/phpstan/phpstan-symfony?

**Comment by dantleech** [comment]
Hi, yes the approach for getting the SF container is the same approach that you would use in PHPUnit etc.  I have done something like this:

```
use App\Kernel;
use PHPUnit\Framework\TestCase;
use Psr\Container\ContainerInterface;

abstract class MyBenchCase
{
    public function container(): ContainerInterface
    {
        $kernel = new Kernel('prod', true);
        $kernel->boot();

        return $kernel->getContainer();
    }
}

/**
 * @BeforeMethods({"setUp"})
 * /
class MyBench {
    public function setUp(): void
    {
        $container = $this->container();
        $this->myService = $container->get(MyService::class);
    } 
}
```

(no extensions required)

**Reply by darthf1** [reply]
Ah I didn't know it was that easy to boot the kernel and get a full container. Many thanks for your reply.

**Reply by WissameMekhilef** [reply]
Hi, I stumbled across this thread while I was trying to fix the same problem.

This solution only works if the service is public, which is not recommended according to the Symfony documentation [here](https://symfony.com/doc/current/service_container.html#public-versus-private-services)

**Reply by DavidGarciaCat** [reply]
I'm facing the same issue here; can't inject private services (but this has been a recurrent issue since Symfony 3, I believe).

Are there any other ways to inject the Symfony's Container or use the Dependency Injection automatically?

**Reply by aszenz** [reply]
I solved this in symfony by using test container service which allows fetching private services.

```php
$container = $this->kernel->getContainer();

$this->container = $container->get('test.service_container'):

$doctrine = $this->container->get('doctrine');
```

Note for this to work the kernel app env must be test.

See:
https://symfony.com/blog/new-in-symfony-4-1-simpler-service-testing
